types/node/util.d.ts

```diff
@@ -1105,6 +1105,197 @@ declare module 'util' {
+    So this helper treats "not definitely present" (i.e., not `extends boolean`) as being "definitely not present", i.e. it should have its default value.
+    This is technically incorrect but is a much nicer UX for the common case.
+    The IfDefaultsTrue version is for things which default to true; the IfDefaultsFalse version is for things which default to false.
+    They differ only in the final item of the if-else chain.
+    */
+    type IfDefaultsTrue<T, IfTrue, IfFalse> = T extends true
+        ? IfTrue
+        : T extends false
+        ? IfFalse
+        : T extends boolean
+        ? IfTrue | IfFalse
+        : IfTrue;
+
+    type IfDefaultsFalse<T, IfTrue, IfFalse> = T extends true
+        ? IfTrue
+        : T extends false
+        ? IfFalse
+        : T extends boolean
+        ? IfTrue | IfFalse
+        : IfFalse;
+
+    type ExtractOptionValue<T extends ParseArgsConfig, O extends ParseArgsOptionConfig> = IfDefaultsTrue<
+        T['strict'],
+        O['type'] extends 'string' ? string : O['type'] extends 'boolean' ? boolean : string | boolean,
+        string | boolean
+    >;
+
+    type ParsedValues<T extends ParseArgsConfig> = (T['options'] extends ParseArgsOptionsConfig
+        ? {
+              -readonly [LongOption in keyof T['options']]: IfDefaultsFalse<
+                  T['options'][LongOption]['multiple'],
+                  undefined | Array<ExtractOptionValue<T, T['options'][LongOption]>>,
+                  undefined | ExtractOptionValue<T, T['options'][LongOption]>
+              >;
+          }
+        : {}) &
+        IfDefaultsTrue<T['strict'], unknown, { [longOption: string]: undefined | string | boolean }>;
+
+    type ParsedPositionals<T extends ParseArgsConfig> = IfDefaultsTrue<
+        T['strict'],
+        IfDefaultsFalse<T['allowPositionals'], string[], []>,
+        IfDefaultsTrue<T['allowPositionals'], string[], []>
+    >;
+
+    type PreciseTokenForOptions<
+        T extends ParseArgsConfig,
+        K extends keyof T['options'] & string,
+        O extends ParseArgsOptionConfig,
+    > = O['type'] extends 'string'
+        ? {
+              kind: 'option';
+              index: number;
+              name: K;
+              rawName: string;
+              value: string;
+              inlineValue: boolean;
+          }
+        : O['type'] extends 'boolean'
+        ? {
+              kind: 'option';
+              index: number;
+              name: K;
+              rawName: string;
+              value: undefined;
+              inlineValue: undefined;
+          }
+        : OptionToken & { name: K };
+
+    type TokenForOptions<
+        T extends ParseArgsConfig,
+        K extends keyof T['options'] = keyof T['options'],
+    > = K extends unknown
+        ? T['options'] extends ParseArgsOptionsConfig
+            ? PreciseTokenForOptions<T, K & string, T['options'][K]>
+            : OptionToken
+        : never;
```

Comment:
Unfortunately doing this prevents the type from being expanded when printed, in my testing - that is, the type for `util.parseArgs(config).tokens[0]` prints as something like

```
| { kind: "positional"; index: number; value: string; }
| { kind: "option-terminator"; index: number; }
| TokenForOptions<{ readonly tokens: true; readonly allowPositionals: true; readonly options: { readonly foo: { readonly type: "string"; }; readonly bar: { readonly type: "boolean"; }; }; }>
```  
(note the presence of the basically-useless "TokenForOptions" in the output) instead of (as currently)

```
| { kind: "positional"; index: number; value: string; }
| { kind: "option-terminator"; index: number; }
| { kind: "option"; index: number; name: "foo"; rawName: string; value: string; inlineValue: boolean; }
| { kind: "option"; index: number; name: "bar"; rawName: string; value: undefined; inlineValue: undefined; }
```

The latter form is much more useful, so I've left this alone for the moment. If there's a way to force the type to get expanded out, or if you think there's some other reason to switch over, I'm happy to adopt this suggestion.

Reply:
I wouldn't depend too much on printback of types, as that has changed in the past, and large types are often truncated when they appear in quick info in the editor.

Reply:
Well, at the very least, the tests in this repo are written against the precise rendering of the types (unfortunately...), so the tests are more readable this way.

Reply:
Its unfortunate that `$ExpectType` is so brittle. I generally use something like https://github.com/esfx/esfx/blob/main/packages/type-model/src/test.ts to test more complex types.